pkg/kubelet/images/image_manager_test.go

```diff
@@ -312,3 +313,71 @@ func TestPullAndListImageWithPodAnnotations(t *testing.T) {
 		assert.Equal(t, expectedAnnotations, image.Spec.Annotations, "image spec annotations")
 	})
 }
+
+func TestMaxParallelImagePullLimit(t *testing.T) {
+	ctx := context.Background()
+	pod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            "test_pod",
+			Namespace:       "test-ns",
+			UID:             "bar",
+			ResourceVersion: "42",
+		}}
+
+	testCase := &pullerTestCase{
+		containerImage: "present_image",
+		testName:       "image present, pull ",
+		policy:         v1.PullAlways,
+		inspectErr:     nil,
+		pullerErr:      nil,
+		qps:            0.0,
+		burst:          0,
+	}
+
+	useSerializedEnv := false
+	maxParallelImagePulls := 5
+	var wg sync.WaitGroup
+
+	puller, fakeClock, fakeRuntime, container := pullerTestEnv(*testCase, useSerializedEnv, maxParallelImagePulls)
+	fakeRuntime.BlockImagePulls = true
+	fakeRuntime.CalledFunctions = nil
+	fakeRuntime.T = t
+	fakeClock.Step(time.Second)
+
+	// First 5 EnsureImageExists should result in runtime calls
+	for i := 0; i < maxParallelImagePulls; i++ {
+		wg.Add(1)
+		go func() {
+			_, _, err := puller.EnsureImageExists(ctx, pod, container, nil, nil)
+			assert.Nil(t, err)
+			wg.Done()
+		}()
+	}
+	time.Sleep(1 * time.Second)
+	fakeRuntime.AssertCallCounts("PullImage", 5)
+
+	// Next two EnsureImageExists should be blocked because maxParallelImagePulls is hit
+	for i := 0; i < 2; i++ {
+		wg.Add(1)
+		go func() {
+			_, _, err := puller.EnsureImageExists(ctx, pod, container, nil, nil)
+			assert.Nil(t, err)
+			wg.Done()
+		}()
+	}
+	time.Sleep(1 * time.Second)
+	fakeRuntime.AssertCallCounts("PullImage", 5)
+
+	// Unblock two image pulls from runtime, and two EnsureImageExists can go through
+	fakeRuntime.UnblockImagePulls(2)
+	time.Sleep(1 * time.Second)
+	fakeRuntime.AssertCallCounts("PullImage", 7)
+
+	// Unblock the remaining 5 image pulls from runtime, and all EnsureImageExists can go through
+	fakeRuntime.UnblockImagePulls(5)
+
+	time.Sleep(1 * time.Second)
+	fakeRuntime.AssertCallCounts("PullImage", 7)
+
+	wg.Wait()
```

Comment:
You should be able to drop the sleep here if you reorder:

```suggestion
	wg.Wait()
	fakeRuntime.AssertCallCounts("PullImage", 7)
```

All the goroutines should be complete by the time `wg.Wait()` returns.